src/libs/ck-libs/ampi/ampi_globals_pie.C

```diff
@@ -0,0 +1,667 @@
+
+#ifndef _GNU_SOURCE
+# define _GNU_SOURCE
+#endif
+#ifndef __USE_GNU
+# define __USE_GNU
+#endif
+
+#include "ampi_funcptr_loader.h"
+#undef atexit
+
+#include <string>
+#include <atomic>
+#include <vector>
+#include <map>
+#include <functional>
+
+#if CMK_HAS_DL_ITERATE_PHDR
+#include <link.h>
+
+#if CMK_HAS_ELF_H
+#include <elf.h>
+#endif
+
+#if CMK_HAS_MMAP
+#include <sys/mman.h>
+#endif
+
+#ifndef PT_GNU_PROPERTY
+#define PT_GNU_PROPERTY 0x6474e553
+#endif
+#endif
+
+
+#define PIEGLOBALS_DEBUG 0
+
+
+struct itemstruct
+{
+  size_t offset;
+  size_t size;
+  int flags;
+};
+
+struct objectstruct
+{
+  std::string name{};
+  char * source = nullptr;
+  size_t size = 0;
+  size_t align = 0;
+  std::vector<itemstruct> items{};
+};
+
+struct pieglobalsstruct
+{
+  ampi_mainstruct mainstruct;
+  SharedObject myexe;
+  ptrdiff_t main_c_diff, main_f_diff;
+
+  size_t numobjects = 0;
+  std::vector<objectstruct> objects{};
+
+  std::vector<std::tuple<ptrdiff_t, uintptr_t, size_t>> segment_allocation_offsets{};
+
+  CmiIsomallocContext heap_context;
+  std::vector<std::tuple<uintptr_t, size_t, size_t>> global_constructor_heap{};
+
+  int nodeleaderrank;
+
+  void dealloc()
+  {
+    global_constructor_heap.clear();
+  }
+};
+
+
+static int phdr_print(struct dl_phdr_info * info, size_t size, void * data)
+{
+  auto & numobjects = *(size_t *)data;
+  if (numobjects > 0)
+  {
+    --numobjects;
+    return 0;
+  }
+
+#if PIEGLOBALS_DEBUG >= 1
+  CmiPrintf("pieglobals> Object - \"%s\" - %d segments\n", info->dlpi_name, info->dlpi_phnum);
+#endif
+
+  for (int j = 0; j < info->dlpi_phnum; j++)
+  {
+    const auto & phdr = info->dlpi_phdr[j];
+    const int p_type = phdr.p_type;
+
+    if (p_type == PT_TLS)
+      CmiError("AMPI> Warning: Use of thread_local detected in pieglobals binary!\n");
```

Comment:
I have successfully built binaries with `-pieglobals -tlsglobals`. However, tlsglobals won't do its job in that mode. The key change that needs to be made is that the size of the TLS segment needs to be determined after pieglobals calls dlopen, instead of during the creation of the CthThread.